### ISSUE TITLE
Improve usability of VPC subnet CIDR distribution

### ISSUE THREAD

**Issue by danielrbradley** [pulumi/pulumi-awsx#1135]
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

### Current challenges

- Unpredictable layout of cidr blocks for subnets due to re-ordering of subnets from their specifications.
- Unable to create layouts with specific ordering (e.g. public before private).
- Unable to create layouts with specific gaps.
- Only using bit-masks are hard to visualize how to specify correctly.
- Difficult to visualize result when only including CIDR block range

### Proposal

1. Add a new `Unused` subnet type. This has no effect in the legacy subnet allocation method.
2. Add `subnetStrategy` field to VPC as enum of `Legacy`, `Auto`, `Exact`. 
	- `Legacy` maintains current complex behaviour.
	- `Auto` lays out in specified order with gaps where required. Either all mask must be specified, or no masks specified - no mixing of auto-allocated and manual is allowed.
	- `Exact` requires the user to specify what every part of the range is used for - using the `Unused` subnet spec type for expected gaps between subnets.

Anyone using the legacy subnet allocation method can definitely migrate to the exact layout method, they can also test migrating to the Auto layout, but it's not guaranteed to be compatible.

In the next major version, we should change the default subnet strategy to `Auto`.

### Example

```typescript
const vpc = new awsx.ec2.Vpc("vpc", {
    cidrBlock: "10.0.0.0/16",
    numberOfAvailabilityZones: 3,
    subnetStrategy: "Exact",
    subnetSpecs: [{
        type: "Public",
        name: "public-ecs-subnet",
        cidrBlocks: ["10.0.0.0/21", "10.0.64.0/21", "10.0.128.0/21"],
    },{
        type: "Unused",
        size: 2048, // same as cidrMask: 21
    },{
        type: "Private",
        name: "private-ecs-subnet",
        cidrMask: 20, // Same as size: 4096
    },{
        type: "Unused",
        size: 8192, // Same as cidrMask: 19
    }]
});
```

### Further work

- Add optional `size` and `cidrBlocks` fields to subnet spec as an alternative to `cidrMask` as a more intuitive way of specifying masks. The `size` must be a power of two (2, 4, 8, 16, 32, 64, 128, etc). The `cidrBlocks` property is an array of strings, which must match the number of AZs being used. Specifying more than one of `cidrMask`, `cidrBlocks` and `size` results in an error.
- Add optional VPC argument for `availabilityZoneCidrMask` to allow the user to not allocate all available space. E.g. if the VPC is /16 and there's initially 2 AZs, we would normally allocate each AZ a /17 mask. However, the user might want to only allocate a /16 mask to give them option to increase the number of AZs in the future to 3 or 4 without recreation.
- Add additional `subnetLayout` output to include each subnet with CIDR block, bitmask, size, usable addresses, wildcard mask, and its host, broadcast, first and last addresses.

### References

- https://github.com/pulumi/pulumi-awsx/issues/951
- https://github.com/pulumi/pulumi-awsx/issues/321
- https://github.com/pulumi/pulumi-awsx/issues/520
- https://github.com/pulumi/pulumi-awsx/issues/982
- https://github.com/pulumi/pulumi-awsx/issues/1152

**Comment by jkodroff** [comment]
Rather than specifying the number of hosts, I would suggest "Exact" should allow the user to specify an explicit CIDR as this removes all abstraction and allows the most user control. (This is also how the TF VPC module does things.) Maybe rename "Exact" to "ExplicitCidrBlock" (or something like that). Options could be named "CidrMask" or "CidrBlock" maybe (and would still be mutually exclusive)?

For auto, part of the current implementation that causes so much trouble is trying to keep all traffic within an AZ, then subdividing from there. Inter-AZ traffic can be quite expensive (would want to @jaxxstorm to confirm how expensive it is in practice), so I don't know if the "Auto" option is needed.

We should test early on to see if we can allow people to get from the current behavior to "Exact". That would be a huge win in allowing users to get out of a sticky situation if they can't alter their VPC today.

**Comment by danielrbradley** [comment]
I've updated the proposal and example to also include a `cidrBlocks` subnet spec option to allow explicit calculation done by the user.

This updated implementation still uses the exact same approach of diving up the space by AZ first - then repeating the subnet pattern into each AZ. 

The "Auto" option is very similar to the "Exact" option except:
1. It doesn't require the explicit declaration of any "Unused" blocks - these can be implicit.
2. All size specifications can be omitted and equal mask will be used (nicer for getting started).

Migrating to this will be much easier if we also implement the `subnetLayout` output which will describe the exact configuration needed to be passed back in for the "Exact" option.

**Comment by ghostsquad** [comment]
I just ran into this issue. Dang... I was hoping there was already a way to solve the problem I had. I want an overall `/16`, subdivided as 3x /18 and 3x /20 (with a /20 remaining being unused).
